docs/types/contexts.md

```diff
@@ -43,59 +47,78 @@ the compiler, and hence can be automatically lifted to aid usability.
+>   `Exception` monadic context. This is related to broken values.
+> 3. Contexts can be combined by using the standard typeset operators, or nested
+>   through repeated uses of `in`.
+
+There is no special syntax for contexts anymore.
+Since [#3828](https://github.com/enso-org/enso/pull/3828) 
+Enso is no longer relaying on a haskelly solution. 
+Rather than that _contexts_ are being manupulated by
+_standard library_ functions grouped around 
+`Standard.Base.Runtime.Context` & co.
+```ruby
+Runtime.Context.Output.with_enabled <|
+    File.new "c:\trash.txt" . delete
+```
+
+There is still the `!` symbol signaling [presence of errors](./errors.md)
+- e.g. _broken values_. However the runtime can handle _broken values_ 
+even without presence of these _exception type signatures_. Thus the
+compiler only verifies the referenced types are valid.
 
 ## Monadic Bind
 
-It is also important to note that Enso has no equivalent to `<-` in Haskell.
-Instead, pure computations are implicitly placed in the `Pure` monadic context,
-and `=` acts to 'peel off' the outermost layer of contexts. As such, this means
-that `=` _always_ acts as `bind`, greatly simplifying how the type-checker has
-to work.
+> [!WARNING]
+> Who knows what `<-` means in Haskell?
+> 
+> It is also important to note that Enso has no equivalent to `<-` in Haskell.
+> Instead, pure computations are implicitly placed in the `Pure` monadic context,
+> and `=` acts to 'peel off' the outermost layer of contexts. As such, this means
+> that `=` _always_ acts as `bind`, greatly simplifying how the type-checker has
+> to work.
 
 ## Context Definitions
 
-Contexts can be defined by users.
-
-> The actionables for this section are:
->
-> - How, what, when and why?
+The supported contexts are `Input`, `Output` and `Dataflow_Stack_Trace` as of
+Enso 2024.5.1 version. Users cannot define their own.
 
 ## Context Lifting
 
-> The actionables for this section are:
+> [!WARNING]
+> Is there any meaning of _lifting_ when there is no special support in type system?
 >
 > - Specify and explain how automated lifting of monadic contexts works.
 > - It depends on the order of `runCtx`
 
 ## Inbuilt Contexts
 
-Enso includes a set of commonly-used monadic contexts as part of `Base`, its
-standard library. These are listed below.
+Enso standard library defines `Input`, `Output` and `Dataflow_Stack_Trace` 
+contects.
 
-> The actionables for this section are:
->
-> - Determine the full set of contexts that Enso should provide by default.
-
-### IO
+### State
 
-> The actionables for this section are:
->
-> - Determine the granularity of IO (it's not one context, but a lot).
-> - Explain how there is no `runIO`, and IO is just run at the program boundary,
->   as well as the impacts of this.
+State acts as a _thread local_ variable (well, it will when #7117 gets fixed): 
+- one can set it up
+- execute some code
+- a code somewhere deep the stack
+- may pick the state up
```

Comment:
We are replacing Haskell-specific metaphors with Java-specific ones 😅 (At least Enso has more to do with Java than Haskell so I guess the direction is right). At least that's what thread local sounds to me - from language semantics point of view it is an implementation detail. From semantics perspective it is just state and what matters how it is propagated - does it depend on lexical scope or dynamic scope? If it's implemented by thread local that perhaps means dynamic scoping.

Maybe we should include examples like:
```
func1 x =
    Context.Output.with_enabled <|
	    closure1 y =
	        [x, y, Context.Output.is_enabled]

main =
    f1 = func1 1
    r = Context.Output.with_disabled <|
        f1 2
    IO.println r
```
prints `[1, 2, False]` because the scope that matters is at execution time and not the lexical scope.

`State` behaves similarly. This has some interesting and non-trivial implications especially for lazy-values.
```
type Lazy_Ref
    Value ~get

main =
    ref = Context.Output.with_enabled <|
        suspended_block =
             Context.Output.is_enabled
        Lazy_Ref.Value suspended_block
    IO.println <| Context.Output.with_disabled <| ref.get
    IO.println <| Context.Output.with_enabled <| ref.get
```
prints `False` twice - because the `suspended_block` is computed upon first access of the lazy variable, within the dynamic scope of that access, not within the scope where it was instantiated (which IMO intuition suggests that `suspended_block` is actually a thing that is computed in the enabled scope upon construction, just the computation itself is deferred).

Reply:
> We are replacing Haskell-specific metaphors with Java-specific ones 😅 (At least Enso has more to do with Java than Haskell so I guess the direction is right). 

Funny. Accurate.

However I don't think there is anything Java specific on [thread local variables](https://en.wikipedia.org/wiki/Thread-local_storage) it is an **operating system concept**. 

> from language semantics point of view it is an implementation detail. 

Uff. I am afraid we are not ready to specify language in such a formal way. Luckily, `State` isn't a language concept (anymore)! `State` is a library function that enables access to [thread local](https://en.wikipedia.org/wiki/Thread-local_storage)-like "variables".

> From semantics perspective it is just state and what matters how it is propagated - does it depend on 
> lexical scope or dynamic scope? If it's implemented by thread local that perhaps means dynamic scoping.
> 
> Maybe we should include examples like:

I am afraid of including such a detailed examples in the documentation because I am afraid they may be incorrect (because of typo, or because of aging). But
- I'd like to link from this documentation to docs for Standard libraries
- I'd like to extract code snippets from real tests - like my [codesnippet4javadoc](https://github.com/jtulach/codesnippet4javadoc/blob/master/README.md) does for Java

Do we have a preferred framework for that? What needs to be done to make it useful? CCing @jdunkerley

Reply:
Reworded in 5ec72a3d1e where I am trying to explain the `State` in terms of _stack_ - future improvements indeed possible.